builld: `sdcard` target feature proposal

commodo opened the pull request:
I need an SD-card image generation logic for Zynq.
While looking at other targets, I found some duplication of code, which I thought would be neat to put into an `sdcard` feature.

The base logic for this was the RPi target.

Tested that the sdcard images get generated for omap, at91, sunxi, mvebu, brcm2708.
Tested that it runs only brcm2708 [brcm2710 subtarget].
Not tested: arc770, archs38 ; but the image generation scripts are similar that I only pointed to the common SD-card generation script.

Not doing [in this patchset]: mxs ; that one seems a bit non-standard than the rest, because there is a vfat and ext4 image generation script ; and the partition layout seems different.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
brcm2708: switch to using `sdcard` feature in target

Re-using the `gen_rpi_sdcard_img.sh` script as base for the SD image
generation.

Add `Build/sdcard-image` command in `image-commands.mk`
All arguments have been wrapped with quotes to ensure positioning if any of
the arguments is to be empty.

Replaced build rules in `target/linux/brcm2708/image/Makefile` to use new
`sdcard-image` rule.

Use generic `CONFIG_TARGET_SDCARD_BOOT_PARTSIZE` instead of target specific
`CONFIG_BRCM2708_SD_BOOT_PARTSIZE` symbol.

Partition alignment default has been reduced to 1024, since that is the
more popular default in other SD card images.  Not sure, if this should be
made configurable per-target.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
build,target: add `sdcard` feature specifier and symbols

targets can add `FEATURES += sdcard` and that will enable the USE_SDCARD
symbol which will define the build rules for building the sdcard image.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
omap: switch to using `sdcard` feature specifier

The `gen_omap_sdcard_img.sh` script is identical to the common
`gen_rpi_sdcard_img.sh`.

Rest of the image logic falls into place to use the common sdcard image
generation logic.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
arc770: use default `gen_sdcard_image.sh` script

The current default `gen_sdcard_image.sh` script is identical to the one
for the arc770 target.  Makes sense to re-use the common one.

Revamping the image generation logic would take a bit more work.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
archs38: use default `gen_sdcard_image.sh` script

The current default `gen_sdcard_image.sh` script is identical to the one
for the arc770 target.  Makes sense to re-use the common one.

Revamping the image generation logic would take a bit more work.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
at91: enable `sdcard` feature for sama5 sub-targets

This will make the number of sectors 63, same as for the rest of the
targets.

Not sure if this is acceptable, or whether some form of configuration
should be added.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
sunxi: add `sdcard` feature support to target

This requires some modifications to the `gen_sdcard_image.sh` script to
allow for an optional (6th) parameter, which is the u-boot image.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo committed:
mvebu: add `sdcard` feature support to target

This replaces the default settings for SD card image generation with the
global defaults of the `sdcard` feature & symbols.

The `UBOOT` variables have been renamed to UBOOT_IMAGE, to be a bit more
explicit.

The SIGNATURE variable is now included in the `get_sdcard_image.sh` script.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo commented:
hmm ; weird ; Github has re-ordered commits a bit ;

nwf commented on scripts/gen_sdcard_image.sh:
```diff
@@ -24,6 +31,3 @@ ROOTFSSIZE="$(($4 / 512))"
 
 dd bs=512 if="$BOOTFS" of="$OUTPUT" seek="$BOOTOFFSET" conv=notrunc
 dd bs=512 if="$ROOTFS" of="$OUTPUT" seek="$ROOTFSOFFSET" conv=notrunc
```
This is nitpicky, but since you're rummaging around in the relevant code...

Since partitions are aligned to 1Mi, the dd block size can be made much bigger; in local testing, this dramatically increased the speed of these dd commands. Either adjust the BOOTOFFSET and ROOTFSOFFSET calculations or don't bother dividing and use oflags=seek_bytes.

commodo replied:
hey, thanks for the note
i'll take a look

tbh, i've been trying to figure out some ideal settings based on what all targets were doing for SD cards [and comparing with what people are saying on the web] ;
i couldn't reach a good conclusion, so i just used what was the more common code

will take a look
thanks

bobafetthotmail replied:
If you are writing in block devices you can use other tools than dd too, skip having to deal with blocksizes and get better performance overall. See this article (and the articles it links as sources) https://unix.stackexchange.com/questions/224277/is-it-better-to-use-cat-dd-pv-or-another-procedure-to-copy-a-cd-dvd#224314

commodo replied:
@nwf 
i compared Linux vs MacOS vs FreeBSD `dd`
`oflag` is only part of Linux's
http://man7.org/linux/man-pages/man1/dd.1.html
https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/dd.1.html

@bobafetthotmail 
thanks for the links

maybe I'll just make the dd blocksize up to 1MB
personally I don't have a strong preference here, since the image sizes I typically used were ~256 MB

commodo replied:
[ was reading up on various stuff about `dd` ]
i'm a little puzzled on the various explanations about `dd` and the rest of the scripts ;
i'd suggest we defer the blocksize stuff for later ;

this version of the script has worked for now; 
the scope of this changeset is common `sdcard` image generation [or start in that direction] ;
i'd like like to leave it at this :)

nwf replied:
That sounds fine to me. :+1:

commodo commented:
removed RFC tag ;

commodo commented:
Changelog V1 -> V2:
* Rebased on newer master

stintel commented:
I very much like the idea. There is too much duplicate code for generating sdcard images and it really needs a cleanup. Thanks for this! Will do a full review later.

stintel commented:
One common remark: when updating copyright, do not simply replace the old year with the current year, but add the current year so it becomes a range. Aside from that, all copyright lines containing OpenWrt.org or LEDE Project are wrong anyway. I would leave the copyright line as is, and only in case you made noteworthy changes, add a copyright line with your name and email address (same as git author / SoB).

commodo commented:
hmm ; i forgot to push a rebase here;
also, due to the issue found here https://github.com/openwrt/openwrt/pull/606#issuecomment-366466480
i may have to investigate if it transmits here too

commodo commented:
Regarding the copyright lines ; I don't feel I added anything new/note-worthy to add myself anywhere in this changeset.
I just moved stuff around to reduce duplication.

I added the Copyright for 2018 in `gen_sdcard_image.sh` ;that file did not have any copyright.
In the PR-view it shows as if I renamed a file.
In the commit, it's this: https://github.com/openwrt/openwrt/pull/687/commits/a0c31d1e075724f2595770bb0b5ff770901d38e8

I probably did not understand what you meant by `all copyright lines containing OpenWrt.org or LEDE Project are wrong anyway`

In any case I can probably remove the Copyright lines from that file or update them.
Your preference.

commodo commented:
I was thinking these last few days of maybe merging `sdcard-boot | sdcard-image`  into `sdcard-image`.
And call `scard-boot` from `sdcard-image`.
Not yet sure about it ; maybe it would make sense later; especially there aren't different `sdcard-boot` image generation stuff for different devices in the same target.

commodo commented:
Changelog V2 -> V3:
* rebased on newer master (8651f1149dc18f6c730ae20f36a91d4eb360a9b0)

stintel commented:
The copyright lines are invalid because you need a signed contract to be able to transfer copyright from you to OpenWrt.org or LEDE Project.

Since you based scripts/gen_sdcard_image.sh on an existing script without copyright info, I would just leave the script as is, without copyright info.

commodo commented:
i see 
that was not obvious to me
thanks for clarifying ;

will re-spin without copyright stuff

commodo commented:
Changelog v3 -> v4:
* rebased on top of newer master (138c76332b7e39b3c4e3018e7792e74645ab0c7a) to resolve conflicts
* removed OpenWrt.org copyright from `gen_sdcard_image.sh`

commodo commented:
will take a look
i think i may need to revisit this, to handle the mvebu sdcard stuff; initially this would justs create a simple FAT32 SD-card; I've noticed that mvebu creates EXT4 SD-card images

commodo committed:
scripts/gen_sdcard_image.sh: add signature from SOURCE_DATE_EPOCH

The number from SOURCE_DATE_EPOCH is padded with zeros at the beginning.
If the number in SOURCE_DATE_EPOCH is invalid, 8 zeros will be added, which
is not ideal, but still ok for build reproduce-ability.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo commented:
Changelog v4 -> v5:
* rebased on newer master ( 56ae9f9b0b1583e02815f0e60bf232947c27c6cf )
* added `SIGNATURE` var in `gen_sdcard_image.sh` ; obtained from `SOURCE_DATE_EPOCH`
* re-applied to `mvebu` target
* tested that all SD card images get generated ; did not run any of them this time ;
* `mxs` still not included ; i took a closer look and there seems to be a `sdimage` utility that is very custom; it's here https://github.com/mhei/fsl-imx-uuc 
  * @mhei any idea if that could be reworked to not be used ? or is the iMX too custom to use a OpenWrt's SD-card generation logic ?

@hauke   please take a look

later, maybe we could improve the `gen_sdcard_image.sh` script ;
it does feel like it could be improved, but atm, the aim is duplication reduction

commodo commented:
Changelog v5 -> v6:
* I just noticed after writing the previous changelog, that the `sama5` subtarget (in at91) had it's `fpu` feature being removed (via commit 2391d6a29e3856f71de844ba2c1307fb6f36747b ) and I was adding it back because I did not resolve correctly a conflict ; I resolved it now

mhei commented:
@commodo: the iMX23/28 internal bootloaders are looking for a small header, this is what the sdimage tool is adding. Maybe a generic openwrt SD card generation script could provide a hook which is not used for others and is overwritten by mxs target?

commodo commented:
managed to test succesfully on a RPi3

commodo commented:
Changelog v6 -> v7:
* rebased on newer master ( 99f10b193e24c59b43382f5e7739bc83853382cf ) - updated `at91` @SandeepSM  - could you test that it works ? I checked that images get generated after my change, but I don't have any Sama5 targets to test

mhei commented:
@commodo: I finally found time to have a closer look at this series: great work - looks good to me. I agree that this should be a starting point for further fine-tuning, e.g. block size and support for mxs platform etc.

commodo commented:
@mhei  - thanks for the review ; it's appreciated :)

tmn505 commented on scripts/gen_sdcard_image.sh:
```diff
@@ -19,16 +13,21 @@ BOOTFSSIZE="$4"
 UBOOT="$6"
 
+[ -z "$SOURCE_DATE_EPOCH" ] || \
+	SIGNATURE="-S $(printf "%.8d" $SOURCE_DATE_EPOCH)"
```
Missing `0x` before signature string, without that `ptgen` assigns random value.
Also this should be handled literally as string, because with current setting it doesn't truncate the provided `$SOURCE_DATE_EPOCH` and `ptgen` takes last eight digits. This breaks booting of ESPRESSObin which relies on the same string but is handled like this:
https://github.com/openwrt/openwrt/blob/dc629d9cf5afc03f67a294078ca1001e831c66c9/target/linux/mvebu/image/Makefile#L16
This takes first eight digits from `$SOURCE_DATE_EPOCH`.

commodo replied:
hmm ; i'll use that;
i tried to run the command separately and it gave me different results ; but i'll update it if you say it works with `$(shell printf "%.8s" $(SOURCE_DATE_EPOCH)) `

tmn505 replied:
You still forgot about `0x`, the line should look like this:
`SIGNATURE="-S 0x$(printf "%.8d" $SOURCE_DATE_EPOCH)"`

commodo replied:
my bad; i misunderstood ; will update

tmn505 replied:
Sorry, You didn't misunderstand, I copy-pasted the old line and added `0x`, the correct one shoul look like this:
`SIGNATURE="-S 0x$(printf "%.8s" $SOURCE_DATE_EPOCH)"`
Sorry again for adding to the confusion.

commodo replied:
no worries;
thanks for the update :)

updated PR

tmn505 commented on include/image-commands.mk:
```diff
@@ -298,3 +298,11 @@ define Build/kernel2minor
+
+define Build/sdcard-image
+	if [ -n "$(UBOOT_IMAGE)" ]; then UBOOT_IMAGE="$(STAGING_DIR_IMAGE)/$(UBOOT_IMAGE)"; fi; \
+	$(SCRIPT_DIR)/gen_sdcard_image.sh "$@" "$@.boot" "$(IMAGE_ROOTFS)" \
```
The command expects bootfs image to have `.boot` suffix, but definition in:
https://github.com/openwrt/openwrt/blob/dc629d9cf5afc03f67a294078ca1001e831c66c9/target/linux/mvebu/image/Makefile#L43
creates bootfs image with `.bootimg` suffix. That produces sdcard image with empty boot partition.

commodo replied:
good catch ; i think renaming `$@.bootimg` => `$@.boot` should be a reasonable solution for now

tmn505 reviewed (changes requested):
Overall looks good, thanks for tackling this.

SandeepSM commented:
I am able to verify this series on at91 targets and looks good to me. Initially for at91 target, creating sdcard image is done only for ext4 or squash file system and not creating it for ubifs by adding checks. however in this series patch, it create sdcard image even for ubifs rootfs and ubifs does not work on block devices.(i.e on sdcard).

commodo commented:
@SandeepSM 
Good point; will limit images to ` (TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS)
`

commodo commented:
@SandeepSM 
I am also a bit puzzled.
Normally, the code that generates SD-card images should not be run by the code that generates ubifs images.
The change on the `at91` target only replaced some parts of the build to use the generic `sdcard` facility ; if ubifs sd-card images are generated with this change, they probably were generated before as well.

commodo commented:
Turns out, I forgot that sdcard images get generated manually; they don't depend on any of the `(TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS)` symbols.

A lot of the images are FAT images [generated with `mkfs.fat` ] ; maybe at some point this could be consolidated; not sure.
The EXT4 images are also generated with `make_ext4fs` .

commodo commented:
@tmn505 thanks for the review; updated PR with your comments

commodo commented:
@tmn505   updated

commodo commented:
Rebased on top of newer master ( 83483ba787e5c972099da1e983e95f396961c7f2 ) to resolve conflicts

stintel commented:
Tested on RPi0W. The image boots, unfortunately I lost my settings after sysupgrade. Would be nice if we can avoid that. Breaking sysupgrade is a really annoying thing to do ;-)

Maybe it's related to the fact that it's writing a new partition table. I remember losing settings before with other devices whenever that message came up. Not sure why it claims the partition layout changed:

```
root@bedroom-rpi0w:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root               503.9M     34.3M    459.4M   7% /
tmpfs                   219.2M    576.0K    218.6M   0% /tmp
/dev/mmcblk0p1          255.7M      9.5M    246.2M   4% /boot
tmpfs                   512.0K         0    512.0K   0% /dev
root@bedroom-rpi0w:~# sysupgrade /tmp/openwrt-brcm2708-bcm2708-rpi-ext4-sdcard.img.gz
Reading partition table from bootdisk...
Reading partition table from image...
Partition layout has changed. Full image will be written.
Saving config files...
Commencing upgrade. All shell sessions will be closed now.
```

After flash:
```
root@bedroom-rpi0w:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root               503.9M     34.6M    459.1M   7% /
tmpfs                   219.2M    440.0K    218.7M   0% /tmp
/dev/mmcblk0p1          255.7M      9.5M    246.3M   4% /boot
tmpfs                   512.0K         0    512.0K   0% /dev
```

stintel commented:
Ok so the partition layout did in fact change (probably because of the different sector size in scripts/gen_sdcard_image.sh?):

Before:
```
Device         Boot  Start     End Sectors  Size Id Type
/dev/mmcblk0p1 *      8192  532479  524288  256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      540672 1589247 1048576  512M 83 Linux
```

After:
```
Device         Boot  Start     End Sectors  Size Id Type
/dev/mmcblk0p1 *      2048  526335  524288  256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      528384 1576959 1048576  512M 83 Linux
```

I guess we should figure out why sysupgrade fails when writing the full image.

commodo commented:
I'll take a look in the next couple of days.
I'm a bit vague on the sysupgrade thing.

commodo commented:
closing & dropping code
doing some cleanup on my github account

nwf commented:
@commodo Thank you for your efforts so far.  Would you mind putting that branch back so someone else can pick up the work?
